lib/server.js

```diff
@@ -163,13 +160,46 @@ class SimplyImitatedSQSHttpServer {
     return this.sqs.items()
   }
 
+  checkWaiters () {
+    var newWaiters = []
+    for (var i = 0; i < this.waiters.length; i++) {
+      var waiter = this.waiters[i]
+      if (
+        waiter.type === 'send' &&
+        this.sendItems >= waiter.count
+      ) {
+        process.nextTick(waiter.cb)
+        continue
+      }
+
+      if (
+        waiter.type === 'pending' &&
+        this.pendingItems === waiter.count
+      ) {
+        process.nextTick(waiter.cb)
+        continue
+      }
+
+      newWaiters.push(waiter)
+    }
+    this.waiters = newWaiters
+  }
+
   waitForMessages (count, cb) {
     const q = this.getQueue()
     if (q.length >= count) {
       return process.nextTick(cb)
     }
 
-    this.waiters.push({ count: count, cb: cb })
+    this.waiters.push({ type: 'send', count: count, cb: cb })
+  }
+
+  waitForFlush (cb) {
+    if (this.pendingItems === 0) {
+      return process.nextTick(cb)
+    }
+
+    this.waiters.push({ type: 'pending', count: 0, cb: cb })
```

Comment:
What about instead of manually tracking the subscribers, we inherit from `EventEmitter` and reuse that logic?

Reply:
because of managing the `count` logic using the event emitter doesn't match cleanly.

Reply:
would https://nodejs.org/api/events.html#events_emitter_listenercount_eventname help?

Reply:
so to get the count of waiters with type send, you'd do `this.listenerCound('send')`

Reply:
count is how many messages to wait for; not related to how many waiters or listeners there are.

Reply:
oooh gotcha!

Reply:
Would still be possible with event emitters but not really an improvement any more:

```js
const onSend = () => cb()
onSend.count = count
this.on('send', onFlush)

// after send
const listeners = this.listeners('send')
for (const listener of listeners) {
  listener.count--
}
```